Fix issue with removing shared processes from process cache

bnham opened the pull request:
#### f27ec28df1b8e1e9b3f4d44f4ecfeeb0f0d0d0fb
<pre>
Fix issue with removing shared processes from process cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=300344">https://bugs.webkit.org/show_bug.cgi?id=300344</a>
<a href="https://rdar.apple.com/162146236">rdar://162146236</a>

Reviewed by Ryosuke Niwa.

When a shared process (see 300964@main) is removed from the process cache, it causes a crash. Fix
this and add a test.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _setCachedProcessLifetimeForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::setCachedProcessLifetimeForTesting):
(WebKit::WebProcessCache::WebProcessCache):
(WebKit::WebProcessCache::addProcessIfPossible):
(WebKit::WebProcessCache::clear):
(WebKit::WebProcessCache::clearAllProcessesForSession):
(WebKit::WebProcessCache::setApplicationIsActive):
(WebKit::WebProcessCache::removeProcess):
(WebKit::WebProcessCache::CachedProcess::create):
(WebKit::WebProcessCache::CachedProcess::CachedProcess):
* Source/WebKit/UIProcess/WebProcessCache.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, SharedProcessWebProcessCacheCanEvict)):

Canonical link: <a href="https://commits.webkit.org/301173@main">https://commits.webkit.org/301173@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fead43e92bc567a6b9331571db985775a1e2c118

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125136 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35542 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131986 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77005 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/17d96a1b-5052-4268-b288-05179fd2a100) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127013 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53366 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95258 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/13ed55b4-7e96-4b66-a373-46b3ef595c74) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128090 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36314 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/111903 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75801 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dab1ab21-6d90-4fea-8c2c-5a0e5b1fc308) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35220 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30068 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75464 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106079 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30296 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134665 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51946 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39736 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103727 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52380 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108124 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103498 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26357 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48846 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27137 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49011 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51838 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57621 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51205 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54561 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52895 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/WebProcessCache.cpp:
```diff
@@ -356,6 +363,17 @@ void WebProcessCache::clearAllProcessesForSession(PAL::SessionID sessionID)
         m_processesPerSite.remove(key);
 
+    HashMap<WebCore::Site, Ref<CachedProcess>> sharedProcessesPerSite;
+    for (auto& pair : m_sharedProcessesPerSite) {
```
I'd call this `entry` or `item` since it's not really a `std::pair`.

bnham committed:
Fix issue with removing shared processes from process cache
https://bugs.webkit.org/show_bug.cgi?id=300344
rdar://162146236

Reviewed by Ryosuke Niwa.

When a shared process (see 300964@main) is removed from the process cache, it causes a crash. Fix
this and add a test.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _setCachedProcessLifetimeForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::setCachedProcessLifetimeForTesting):
(WebKit::WebProcessCache::WebProcessCache):
(WebKit::WebProcessCache::addProcessIfPossible):
(WebKit::WebProcessCache::clear):
(WebKit::WebProcessCache::clearAllProcessesForSession):
(WebKit::WebProcessCache::setApplicationIsActive):
(WebKit::WebProcessCache::removeProcess):
(WebKit::WebProcessCache::CachedProcess::create):
(WebKit::WebProcessCache::CachedProcess::CachedProcess):
* Source/WebKit/UIProcess/WebProcessCache.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, SharedProcessWebProcessCacheCanEvict)):

Canonical link: https://commits.webkit.org/301173@main
